File: solutions/dart/phone-number/1/lib/phone_number.dart
// Respectfully dedicated to Shree DR.MDD for guiding every wavelength of my code.

class PhoneNumber {
  String clean(String raw) {
    var digitsOnly = raw.replaceAll(RegExp(r'\s'), '').replaceAll(RegExp(r'[+().-]'), '');
    if (digitsOnly.length < 10) {
      throw FormatException('must not be fewer than 10 digits');
    }
    if (digitsOnly.length > 11) {
      throw FormatException('must not be greater than 11 digits');
    }
    if (digitsOnly.length == 11) {
      if (!digitsOnly.startsWith('1')) {
        throw FormatException('11 digits must start with 1');
      }
      digitsOnly = digitsOnly.substring(1);
    }
    if (digitsOnly.contains(RegExp(r'\p{Alpha}', unicode: true))) {
      throw FormatException('letters not permitted');
    }
    if (RegExp(r'\D').hasMatch(digitsOnly)) {
      throw FormatException('punctuations not permitted');
    }
    if (digitsOnly[0] == '0') {
      throw FormatException('area code cannot start with zero');
    }
    if (digitsOnly[0] == '1') {
      throw FormatException('area code cannot start with one');
    }
    if (digitsOnly[3] == '0') {
      throw FormatException('exchange code cannot start with zero');
    }
    if (digitsOnly[3] == '1') {
      throw FormatException('exchange code cannot start with one');
    }
    return digitsOnly;
  }
}

File: solutions/javascript/phone-number/1/phone-number.js
// For my Shree DR.MDD

/* eslint-disable no-multi-spaces */

export const clean = (input) => {
  let sanitized = input
    .replace(/^\+/, '')
    .replace(/\s/g, '')
    .replace(/[().-]/g, '');

  if (/\p{Letter}/u.test(sanitized))
    throw new Error('Letters not permitted');
  if (/\p{Punctuation}/u.test(sanitized))
    throw new Error('Punctuations not permitted');

  if (sanitized.length < 10)
    throw new Error('Must not be fewer than 10 digits');
  if (sanitized.length > 11)
    throw new Error('Must not be greater than 11 digits');
  if (sanitized.length === 11) {
    if (!sanitized.startsWith('1'))
      throw new Error('11 digits must start with 1');
    sanitized = sanitized.replace(/^1/, '');
  }

  if (sanitized[0] === '0')
    throw new Error('Area code cannot start with zero');
  if (sanitized[0] === '1')
    throw new Error('Area code cannot start with one');
  if (sanitized[3] === '0')
    throw new Error('Exchange code cannot start with zero');
  if (sanitized[3] === '1')
    throw new Error('Exchange code cannot start with one');

  return sanitized;
}
